tests/unit/test_charm.py

```diff
@@ -325,3 +349,283 @@ def test_given_environment_information_available_all_relations_created_when_pebb
+            relation_name=SDCORE_MANAGEMENT_RELATION_NAME,
+            remote_app=TEST_SDCORE_MANAGEMENT_PROVIDER_APP_NAME,
+        )
+        self.harness.update_relation_data(
+            relation_id=sdcore_management_relation_id,
+            app_or_unit=TEST_SDCORE_MANAGEMENT_PROVIDER_APP_NAME,
+            key_values={"management_url": "http://10.0.0.1:5000"},
+        )
+        self.harness.container_pebble_ready("nms")
+        patch_push.assert_called_with(
+            path=TEST_UPF_CONFIG_PATH, source=json.dumps(expected_upf_config)
+        )
+
+    @patch("ops.model.Container.push")
+    @patch("ops.model.Container.exists")
+    @patch("ops.model.Container.pull")
+    def test_given_gnb_config_already_pushed_and_content_changes_when_pebble_ready_then_gnb_config_is_pushed(  # noqa: E501
+        self,
+        patch_pull,
+        patch_exists,
+        patch_push,
+    ):
+        existing_gnb_config = [
+            {
+                "name": "some.gnb.name",
+                "tac": "1234",
+            }
+        ]
+        expected_gnb_config = [
+            {
+                "name": "some.gnb.name",
+                "tac": "1234",
+            },
+            {
+                "name": "some.other.gnb.name",
+                "tac": "4567",
+            },
+        ]
+        patch_exists.return_value = True
+        patch_pull.return_value = StringIO(json.dumps(existing_gnb_config))
+
+        gnb_identity_relation_1_id = self.harness.add_relation(
+            relation_name=GNB_IDENTITY_RELATION_NAME,
+            remote_app=TEST_GNB_IDENTITY_PROVIDER_APP_NAME,
+        )
+        self.harness.update_relation_data(
+            relation_id=gnb_identity_relation_1_id,
+            app_or_unit=TEST_GNB_IDENTITY_PROVIDER_APP_NAME,
+            key_values={"gnb_name": "some.gnb.name", "tac": "1234"},
+        )
+        gnb_identity_relation_2_id = self.harness.add_relation(
+            relation_name=GNB_IDENTITY_RELATION_NAME,
+            remote_app=TEST_GNB_IDENTITY_PROVIDER_APP_NAME,
+        )
+        self.harness.update_relation_data(
+            relation_id=gnb_identity_relation_2_id,
+            app_or_unit=TEST_GNB_IDENTITY_PROVIDER_APP_NAME,
+            key_values={"gnb_name": "some.other.gnb.name", "tac": "4567"},
+        )
+        sdcore_management_relation_id = self.harness.add_relation(
+            relation_name=SDCORE_MANAGEMENT_RELATION_NAME,
+            remote_app=TEST_SDCORE_MANAGEMENT_PROVIDER_APP_NAME,
+        )
+        self.harness.update_relation_data(
+            relation_id=sdcore_management_relation_id,
+            app_or_unit=TEST_SDCORE_MANAGEMENT_PROVIDER_APP_NAME,
+            key_values={"management_url": "http://10.0.0.1:5000"},
+        )
+        self.harness.container_pebble_ready("nms")
+        patch_push.assert_called_with(
+            path=TEST_GNB_CONFIG_PATH, source=json.dumps(expected_gnb_config)
+        )
```

Comment:
test cases where the content in the n4 or the gnb_identity relations are incorrect are missing

Reply:
Added